pkg/ocihook/ocihook.go

```diff
@@ -194,17 +202,38 @@ func getCNINamespaceOpts(opts *handlerOpts) ([]cni.NamespaceOpts, error) {
 		if !rootlessutil.IsRootlessChild() {
 			return []cni.NamespaceOpts{cni.WithCapabilityPortMap(opts.ports)}, nil
 		}
+		var (
+			childIP                            net.IP
+			portDriverDisallowsLoopbackChildIP bool
+		)
+		info, err := opts.rootlessKitClient.Info(context.TODO())
+		if err != nil {
+			logrus.WithError(err).Warn("cannot call RootlessKit Info API, make sure you have RootlessKit v0.14.1 or later")
+		} else {
+			childIP = info.NetworkDriver.ChildIP
+			portDriverDisallowsLoopbackChildIP = info.PortDriver.DisallowLoopbackChildIP // true for slirp4netns port driver
+		}
 		// For rootless, we need to modify the hostIP that is not bindable in the child namespace.
 		// https: //github.com/containerd/nerdctl/issues/88
 		//
 		// We must NOT modify opts.ports here, because we use the unmodified opts.ports for
 		// interaction with RootlessKit API.
 		ports := make([]cni.PortMapping, len(opts.ports))
 		for i, p := range opts.ports {
-			if hostIP := net.ParseIP(p.HostIP); hostIP != nil {
+			if hostIP := net.ParseIP(p.HostIP); hostIP != nil && !hostIP.IsUnspecified() {
 				// loopback address is always bindable in the child namespace, but other addresses are unlikely.
 				if !hostIP.IsLoopback() {
-					p.HostIP = "127.0.0.1"
+					if childIP != nil && childIP.Equal(hostIP) {
+						// this is fine
+					} else {
+						if portDriverDisallowsLoopbackChildIP {
+							p.HostIP = childIP.String()
```

Comment:
Is it OK that `childIP == nil` enters this case?

Reply:
Can’t be nil when portDriverDisallowsLoopbackChildIP is true